### PR TITLE
K8SPXC-627 Add logic for choosing host with oldest binlogs

### PR DESCRIPTION
**Pull request by nonemax**
[K8SPXC-627](https://jira.percona.com/browse/K8SPXC-627) [&#10088;?&#10089;](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

When pod from where collector start uploading binlogs is failed collector should connect to pod with the oldest binlogs in cluster. We need to be sure that we didn't have gap between uploaded files

### CODE REVIEW

File: cmd/pitr/pxc/pxc.go [hunk 1]
```diff
@@ -187,12 +208,80 @@ func GetPXCLastHost(pxcServiceName string) (string, error) {
...
+				db.Close()
+				continue
+			}
+			if len(list) == 0 {
+				log.Printf("ERROR: get binlog list for host %s: no binlogs found", nodeArr[0])
+				db.Close()
+				continue
+			}
+			var binlogTime int64
+			for _, binlogName := range list {
```
Comment (delgod) [inline cmd/pitr/pxc/pxc.go:247]: can we just take one the latest item from the list? and not process all binlogs on all servers

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| big-data | passed |
| haproxy | passed |
| scaling | passed |
| upgrade-haproxy | passed |
| recreate | passed |
| self-healing | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| smart-update | passed |
| operator-self-healing | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| pitr | passed |
| users | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| scheduled-backup | failed |

File: cmd/pitr/pxc/pxc.go [hunk 2]
```diff
@@ -187,12 +208,80 @@ func GetPXCLastHost(pxcServiceName string) (string, error) {
...
+		return "", errors.Wrap(err, "get peer-list output")
+	}
+	nodes := strings.Split(string(out), "node:")
+
+	var oldestHost string
+	var oldestTS int64
+	for _, node := range nodes {
+		if strings.Contains(node, "wsrep_ready:ON:wsrep_connected:ON:wsrep_local_state_comment:Synced:wsrep_cluster_status:Primary") {
+			nodeArr := strings.Split(node, ":")
+			db, err := NewPXC(nodeArr[0], user, pass)
```
Comment (qJkee) [inline cmd/pitr/pxc/pxc.go:230]: we have a lot of continue there
to make it more `clear` i think it's worth to put `db.Close()` to defer after opening connection

File: cmd/pitr/pxc/pxc.go [hunk 3]
```diff
@@ -187,12 +208,80 @@ func GetPXCLastHost(pxcServiceName string) (string, error) {
...
 	if len(lastHost) == 0 {
-		return "", errors.New("cant find host")
+		return "", errors.New("can't find host")
 	}
 
 	return lastHost, nil
 }
 
+func GetPXCOldestBinlogHost(pxcServiceName, user, pass string) (string, error) {
+	cmd := exec.Command("peer-list", "-on-start=/usr/bin/get-pxc-state", "-service="+pxcServiceName)
```
Comment (qJkee) [inline cmd/pitr/pxc/pxc.go:218]: can we move this part to separate function?
We already using the same code in 2 places

File: cmd/pitr/pxc/pxc.go [hunk 4]
```diff
@@ -187,12 +208,80 @@ func GetPXCLastHost(pxcServiceName string) (string, error) {
...
+					continue
+				}
+			}
+			if binlogTime == 0 {
+				log.Printf("ERROR: get binlog oldest timestamp for host %s: no binlogs timestamp found", nodeArr[0])
+				db.Close()
+				continue
+			}
+
+			if oldestTS > 0 && binlogTime < oldestTS {
```
Comment (qJkee) [inline cmd/pitr/pxc/pxc.go:265]: more common way there is to set up `oldestTS` to max value
We can initialize it with `oldestTS := math.MaxInt64`, and then just check if `binlogTime` is less that it

File: cmd/pitr/pxc/pxc.go [hunk 5]
```diff
@@ -187,12 +208,80 @@ func GetPXCLastHost(pxcServiceName string) (string, error) {
...
+				oldestHost = nodeArr[0]
+				oldestTS = binlogTime
+			}
+
+			if len(oldestHost) == 0 {
```
Comment (qJkee) [inline cmd/pitr/pxc/pxc.go:270]: we can move this logic to the `if` statement above

File: cmd/pitr/pxc/pxc.go [hunk 6]
```diff
@@ -105,6 +107,25 @@ func (p *PXC) GetBinLogList() ([]Binlog, error) {
 	return binlogs, nil
 }
 
+// GetBinLogList return binary log files list
+func (p *PXC) GetBinLogNamesList() ([]string, error) {
+	rows, err := p.db.Query("SHOW BINARY LOGS")
```
Comment (blez) [inline cmd/pitr/pxc/pxc.go:112]: We should do defer rows.Close() to release connection in case of error

File: cmd/pitr/pxc/pxc.go [hunk 7]
```diff
@@ -187,12 +208,80 @@ func GetPXCLastHost(pxcServiceName string) (string, error) {
...
+
+	var oldestHost string
+	var oldestTS int64
+	for _, node := range nodes {
+		if strings.Contains(node, "wsrep_ready:ON:wsrep_connected:ON:wsrep_local_state_comment:Synced:wsrep_cluster_status:Primary") {
+			nodeArr := strings.Split(node, ":")
+			db, err := NewPXC(nodeArr[0], user, pass)
+			if err != nil {
+				log.Printf("ERROR: creating connection for host %s: %v", nodeArr[0], err)
+				continue
```
Comment (blez) [inline cmd/pitr/pxc/pxc.go:233]: To get rid of all these `continue` it's better to move this logic to separate func and return just error from it.

Comment (nonemax) [reply cmd/pitr/pxc/pxc.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/761/commits/f88ca5e04a895030dc7ef74f1397cecb900c20a1

File: cmd/pitr/pxc/pxc.go (hunk 6 above)
Comment (nonemax) [reply cmd/pitr/pxc/pxc.go:112]: we have Close() method in defer

File: cmd/pitr/pxc/pxc.go (hunk 5 above)
Comment (nonemax) [reply cmd/pitr/pxc/pxc.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/761/commits/f88ca5e04a895030dc7ef74f1397cecb900c20a1

File: cmd/pitr/pxc/pxc.go (hunk 4 above)
Comment (nonemax) [reply cmd/pitr/pxc/pxc.go]: I think we shouldn't do anything with oldestTS before we actually get any data

File: cmd/pitr/pxc/pxc.go (hunk 3 above)
Comment (nonemax) [reply cmd/pitr/pxc/pxc.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/761/commits/ea34dccb63d118b921baf514be8d5993d1a14184

File: cmd/pitr/pxc/pxc.go (hunk 2 above)
Comment (nonemax) [reply cmd/pitr/pxc/pxc.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/761/commits/f88ca5e04a895030dc7ef74f1397cecb900c20a1

File: cmd/pitr/pxc/pxc.go [hunk 8]
```diff
@@ -191,13 +191,20 @@ func (p *PXC) IsGTIDSubset(subSet, gtidSet string) (bool, error) {
...
 	cmd := exec.Command("peer-list", "-on-start=/usr/bin/get-pxc-state", "-service="+pxcServiceName)
 	out, err := cmd.CombinedOutput()
 	if err != nil {
-		return "", errors.Wrap(err, "get peer-list output")
+		return nil, errors.Wrap(err, "get peer-list output")
+	}
+	return strings.Split(string(out), "node:"), nil
+}
+
+func GetPXCLastHost(pxcServiceName string) (string, error) {
```
Comment (delgod) [inline cmd/pitr/pxc/pxc.go:203]:
```suggestion
func GetPXCFirstHost(pxcServiceName string) (string, error) {
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: cmd/pitr/pxc/pxc.go (hunk 6 above)
Comment (blez) [reply cmd/pitr/pxc/pxc.go:112]: As I understand it closes whole db. I'm talking about closing one of connection from pool.

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| scaling | passed |
| upgrade-haproxy | passed |
| haproxy | passed |
| storage | passed |
| recreate | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| smart-update | passed |
| demand-backup | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| self-healing-advanced | passed |
| proxysql-sidecar-res-limits | passed |
| pitr | passed |
| users | passed |
| upgrade-consistency | passed |
| operator-self-healing | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| scheduled-backup | failed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| storage | passed |
| scaling | passed |
| recreate | passed |
| big-data | passed |
| haproxy | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| smart-update | passed |
| demand-backup | passed |
| affinity | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| pitr | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| scheduled-backup | failed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| storage | passed |
| upgrade-haproxy | passed |
| recreate | passed |
| big-data | passed |
| haproxy | passed |
| self-healing | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| security-context | failed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| affinity | passed |
| one-pod | passed |
| smart-update | passed |
| auto-tuning | passed |
| operator-self-healing | passed |
| proxysql-sidecar-res-limits | passed |
| pitr | passed |
| users | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| upgrade-consistency | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| scheduled-backup | failed |

**Review by delgod — APPROVED** [review]

File: cmd/pitr/pxc/pxc.go (hunk 6 above)
Comment (nonemax) [reply cmd/pitr/pxc/pxc.go:112]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/761/commits/76be207197315bc5a9af18272da5705c1660b9e4

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| scaling | passed |
| scaling-proxysql | passed |
| security-context | passed |
| haproxy | passed |
| init-deploy | passed |
| big-data | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| storage | passed |
| self-healing | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| scheduled-backup | failed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Review by blez — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| big-data | passed |
| upgrade-haproxy | passed |
| storage | passed |
| haproxy | passed |
| recreate | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| smart-update | failed |
| affinity | passed |
| demand-backup | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| pitr | passed |
| users | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| scheduled-backup | passed |
| validation-hook | passed |
| upgrade-consistency | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| big-data | failed |
| scaling | passed |
| upgrade-haproxy | passed |
| storage | passed |
| recreate | failed |
| init-deploy | failed |
| self-healing | failed |
| scaling-proxysql | passed |
| security-context | failed |
| upgrade-proxysql | failed |
| restore-to-encrypted-cluster | failed |
| limits | failed |
| self-healing-advanced | failed |
| demand-backup | failed |
| monitoring-2-0 | failed |
| pitr | passed |
| operator-self-healing | failed |
| demand-backup-encrypted-with-tls | passed |
| scheduled-backup | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

**Review by delgod — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| haproxy | passed |
| big-data | passed |
| storage | passed |
| scaling | passed |
| upgrade-haproxy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| self-healing-advanced | passed |
| upgrade-proxysql | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| smart-update | passed |
| demand-backup | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | failed |
| pitr | passed |
| users | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | failed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| storage | passed |
| affinity | passed |
| self-healing | passed |
| one-pod | passed |
| upgrade-haproxy | passed |
| auto-tuning | passed |
| self-healing-advanced | passed |
| proxysql-sidecar-res-limits | passed |
| operator-self-healing | passed |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| scheduled-backup | passed |

**Pull request merged**